### PR TITLE
Timeslot

### LINKED ISSUE #64: Implementation and Testing of TimeSlot class
- [x] Create a new branch for TimeSlot class 
- [x] Change from "to do" to "in progress"
- [x] Create service skeleton 
- [x] Create controller skeleton
- [x] Create DTO skeleton
- [x] Create test class called TestTimeSlotService under src/test/java ca.mcgill.ecse321.repairshop.service package
- [x] Write Service tests

When a task is done: 💯 
_Don't forget to check the checkboxes + add a comment before closing an issue_
_Feel free to create more tasks_

**Links:**
[Tutorial notes](https://mcgill-ecse321-fall2020.github.io/ecse321-tutorial-notes/#_creating_restful_web_services_in_spring)
[Tutorial 6 - REST API](https://lrs.mcgill.ca/embed?rid=6b0fbc0e-223f-4599-9478-bcff69d42270&d2lhost=mycourses2.mcgill.ca)

### PR DESCRIPTION
**Pull request by JDziewonski98**
getall, getbyid, create logic all tested

### CODE REVIEW

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java [hunk 1]
```diff
@@ -1,4 +1,97 @@
...
+        timeslotDto.setEndDateTime(END_TIME);
+        try {
+            TimeSlotDto foundByID = timeSlotService.getTimeslotByID(1L);
+            assertEquals(foundByID.getEndDateTime(), END_TIME);
+            assertEquals(foundByID.getStartDateTime(), START_TIME);
+        } catch (Exception e) {
+            fail(e.getMessage());
+        }
+    }
+
```
Comment (sandyl289) [inline RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java:91]: I think we are also required to test invalid inputs. I tried to write a test, but it failed.

The test I wrote:
```
@Test
  public void testGetTimeslotNonExistent() {
        String error = null;
        Long nonExistentTimeSlotID = 9999l;

        try {
            timeSlotService.getTimeslotByID(nonExistentTimeSlotID);


        } catch (Exception e) {
            error = e.getMessage();
        }
        assertEquals("Timeslot not found", error);
    }
```

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/TimeSlotController.java [hunk 2]
```diff
@@ -17,8 +15,25 @@
...
+    /**
+     * POST to create a timeslot
+     *
+     * @param timeslot DTO with start time and end time
+     * @return HTTP status code
+     */
+    @PostMapping("/create")
+    public ResponseEntity<?> create(@RequestBody TimeSlotDto timeslot) {
+        try {
+            return new ResponseEntity<>(timeSlotService.createTimeslot(timeslot), HttpStatus.OK);
```
Comment (sandyl289) [inline RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/TimeSlotController.java]: Just a question: When do we need to use `HttpsStatus.CREATED`?

**Review by sandyl289 — CHANGES_REQUESTED** [review]
I think we should add one more test to test an invalid input for GetTimeslot(), other than that, everything looks good!

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java (hunk 1 above)
Comment (sandyl289) [reply RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java:91]: Error I get when I add the test I wrote (`testGetTimeslotNonExistent()`)<br/>
image

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java [hunk 3]
```diff
@@ -1,4 +1,97 @@
...
+    @Test
+    public void testGetAllTimeslot() {
+        List<TimeSlotDto> timeslots = timeSlotService.getAllTimeslots();
+        assertEquals(1, timeslots.size()); // List contains a single element.
+        TimeSlotDto timeslot = (TimeSlotDto) timeslots.toArray()[0];
+        assertEquals(ID, timeslot.getID()); // List contains the only comment.
```
Comment (sandyl289) [inline RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java]:
```suggestion
        assertEquals(ID, timeslot.getID()); 
```

File: RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java (hunk 1 above)
Comment (T-Syme) [reply RepairShop-Backend/src/test/java/ca/mcgill/ecse321/repairshop/service/TestTimeSlotService.java:91]: It looks like it's caused by the Optional<TimeSlot> part
image

**Review by T-Syme — COMMENTED** [review]
I agree with Sandy's suggestions. Apart from those, all looks good to me

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/TimeSlotController.java (hunk 2 above)
Comment (T-Syme) [reply RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/TimeSlotController.java]: I'm not sure... If anywhere, this seems like the right place from this (although OK definitely fits too):
image

**Review by Shaswata79 — APPROVED** [review]

File: RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/TimeSlotController.java (hunk 2 above)
Comment (JDziewonski98) [reply RepairShop-Backend/src/main/java/ca/mcgill/ecse321/repairshop/controller/TimeSlotController.java]: yeah this would be the right time to do created

**Pull request merged**